wallet/src/key_chain.rs

```diff
@@ -0,0 +1,186 @@
+use crypto::key::hdkd::derivable::{Derivable, DerivationError};
+use crypto::key::hdkd::derivation_path::DerivationPath;
+use std::str::FromStr;
+use std::sync::Arc;
+use storage::Backend;
+use wallet_storage::{Store, WalletStorageRead};
+use zeroize::Zeroize;
+
+const KEY_KIND: ExtendedKeyKind = ExtendedKeyKind::Secp256k1Schnorr;
+const LOOKAHEAD_SIZE: u16 = 100;
+
+/// KeyChain errors
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum KeyChainError {
+    #[error("Bip39 error: {0}")]
+    Bip39(bip39::Error),
+    #[error("Key derivation error: {0}")]
+    Derivation(#[from] DerivationError),
+    #[error("Address error: {0}")]
+    Address(#[from] AddressError),
+}
+
+/// Result type used for the key chain
+type KeyChainResult<T> = Result<T, KeyChainError>;
+
+/// The usage purpose of a key i.e. if it is for receiving funds or for change
+pub enum KeyPurpose {
+    /// This is for addresses created for receiving funds that are given to the user
+    ReceiveFunds = 0,
+    /// This is for the internal usage of the wallet when creating change output for a transaction
+    Change = 1,
+}
+
+#[allow(dead_code)] // TODO remove
+/// This key chain contains a pool of pre-generated keys and addresses for the usage in a wallet
+pub struct KeyChain<B: Backend> {
+    /// The specific chain this KeyChain is based on, this will affect the address format
+    chain_config: Arc<ChainConfig>,
+
+    /// The master key of this key chain from where all the keys are derived from
+    // TODO implement encryption
+    master_key: ExtendedPrivateKey,
+
+    /// The database connection
+    // TODO Consider if this should be an Option
+    db: Arc<Store<B>>,
+
+    /// The number of addresses to pre-generate from the master key
+    lookahead_size: u16,
+}
+#[allow(dead_code)] // TODO remove
+impl<B: Backend> KeyChain<B> {
+    pub fn new_from_mnemonic(
+        chain_config: Arc<ChainConfig>,
+        mnemonic_str: &str,
+        passphrase: Option<&str>,
+        db: Arc<Store<B>>,
+    ) -> KeyChainResult<Self> {
+        // TODO use Mnemonic::parse when bip39 is configured with "std"
+        let mnemonic =
+            bip39::Mnemonic::parse_normalized(mnemonic_str).map_err(KeyChainError::Bip39)?;
+        // TODO use Mnemonic::to_seed when bip39 is configured with "std"
+        let mut seed = mnemonic.to_seed_normalized(passphrase.unwrap_or(""));
+        let master_key = ExtendedPrivateKey::new_master(&seed, KEY_KIND)?;
+        // TODO(SECURITY) erase mnemonic
+        seed.zeroize();
+
+        Ok(KeyChain {
+            chain_config,
+            master_key,
+            db,
+            lookahead_size: LOOKAHEAD_SIZE,
+        })
+    }
+
+    pub fn load_key_chain(db: Arc<Store<B>>) -> KeyChainResult<Self> {
+        // TODO remove this
+        let _ = db.get_storage_version().expect("This should work?");
+        // TODO implement loading from database
+        Ok(KeyChain {
```

Comment:
To encrypt things you need to store, please use `chacha20poly1305`. It's already done in the crypto trait.

Reply:
Thanks, I will in a new PR.